### ISSUE TITLE
[Looking for suggestions] How to start looking at xla code?

### ISSUE THREAD

**Issue by Crysitna** [pytorch/xla#943]
not really an issue, it's just I met so many random errors from stack-trace-back or other source which I barely understand, and the model training is quite slow and I cannot spot which tensors got various shape at different training and are re-allocated to cpu, so I'm thinking if looking into the `csrc` would help. Yet not quite sure if I wanna find out the reason of the slow training where shall I get started with? thanks so much!

**Comment by dlibenzi** [comment]
I would start setting:
```Shell
export XLA_METRICS_FILE=/PATH/TO/METRICS/FILE
```

The file is always appended, so remove after each run if appending across runs is not what you want.
Or just:
```Python
print(torch_xla._XLAC._xla_metrics_report())
```

Metrics give us/you a first hint of where slowness resides.

**Comment by Crysitna** [comment]
hi thanks for the suggestions that is a good start point. I've tried to look at the `print(torch_xla._XLAC._xla_metrics_report())` also, but u see even if the metrics time are given it's quite hard to find out the exact line of code which causes the slowness and how to speed up.

**Comment by dlibenzi** [comment]
The metrics report is a starting point ;)
You can learn from the replies that are given to certain patterns within the report, how to diagnose further.
Gaining experience on debugging, will also progressively give you insights on the source code.
You can start digging in, but we do not have a book for it, so the learning curve might be a bit steep.

**Comment by Crysitna** [comment]
yeah i was always looking for the book...thanks for telling this tho, will keep look at it.

**Comment by nosound2** [comment]
@Crysitna , there is a short article that I wrote on my experience so far, may be will be of help to you. @dlibenzi if you can please take a look and let me know if I wrote some nonsense!

https://medium.com/@zaharchikishev/running-pytorch-on-tpu-a-bag-of-tricks-b6d0130bddd4?source=friends_link&sk=695f9605874c7306837645ff814e7fc9

**Comment by dlibenzi** [comment]
@nosound2 
Why can't DataParallel run within Jupyter notebook?

Also, can you describe the setup you used for your benchmark?
Was it single core or multi-core?
Where did you get your 110 rate?

**Comment by nosound2** [comment]
Thank you for your feedback!

It is multi-core, and now as single-core is working for me as well, I will include it too in the comparison. 

> Also, can you describe the setup you used for your benchmark?

In that question, what exactly do you think is not clear in the setup description?

Do you think 110 is the expected rate or TPU is under-utilized? I am not sure what is my bottleneck. I tried to `print(torch_xla._XLAC._xla_metrics_report())` but I didn't understand how to make conclusions from that log, a lot of numbers that I didn't know what to do with them.

> If you can link the part of your model which required moving to CPU, we might be able to advise in order to keep it on TPU.

That would be great, attaching my loss function below. There are two implementations, with tensors (slow) and with numpy (fast)

```
    def forward(self, input, target):
        
        NUMPY_LOSS = True
        
        if NUMPY_LOSS:
            input_cpu = input.cpu().detach().numpy()
            target_cpu = target.cpu().numpy()
            
            input_test = input_cpu[target_cpu >= 2000,:]
            target_test = target_cpu[target_cpu >= 2000]
            
            for i in range(input_test.shape[0]):
                input_test[i, sirna_sets != (target_test[i] - 2000)] = 0
            input_test = input_test / np.repeat(input_test.sum(1)[:, np.newaxis], classes, axis=1)

            target_enc = np.zeros_like(input_cpu)
            target_enc[target_cpu < 2000, target_cpu[target_cpu < 2000]] = 1
            target_enc[target_cpu >= 2000, :] = input_test

            loss = - torch.from_numpy(target_enc).to(device=input.device) * torch.log(input + 1e-8)
        else:
            target_test = input[target >= 2000, :].clone().detach()
            sirna_sets_tpu = sirna_sets.to(device=input.device)
            for i in range(target_test.shape[0]):
                target_test[i, sirna_sets_tpu != (target[target >= 2000][i] - 2000)] = 0
            target_test = target_test / target_test.sum(1).repeat((classes,1)).transpose(1,0)
            target_enc = torch.zeros_like(input, device=input.device)
            target_enc[target < 2000, target[target < 2000]] = 1
            target_enc[target >= 2000, :] = target_test
            loss = - target_enc * torch.log(input + 1e-8)
        
        loss = loss.sum(dim=1)
        return loss.mean()
```

it is a regular log likelihood loss, but if the target is equal or more than 2000 I want to take input as a target, but with some probabilities zeroed out (according to static sirna_sets vector). input here is coming from a softmax layer.

**Comment by nosound2** [comment]
> Why can't DataParallel run within Jupyter notebook?

Regarding the jupyter notebook, I am able to run it now, it crashed for me a few times and then started to run OK. It was also giving errors before and killed my kernel several times. Same for running the MNIST example that I moved to a notebook as it is. I updated the article accordingly.

**Comment by dlibenzi** [comment]
How did you measure your 110 rate?
Using our `RateTracker`?

We usually compare 1 TPU v3-8 with 4 V100, performance wise.

The metrics report ... you can post here and ask guidance.

WRT you TPU loss, I'd definitely cache the masks (target > 2000, etc...).
Can you use `torch.where()` instead of open coded loops?
You are also sending `sirna_sets` to TPU every time, which you can probably cache.

**Comment by nosound2** [comment]
Thanks for the valuable feedback. I measured rate 110 by manually measuring the running time. Does `RateTracker` provide any additional insights? I saw it was used in the MNIST example, but I could easily calculate rate myself.

**Comment by dlibenzi** [comment]
Nothing special about `RateTracker`, but for N cores you need to sum all the rate of `xla:*` devices at each step.
IOW, each `RateTracker` reports the per-core rate.
